### PR TITLE
feat(execution): extended-hours execution-layer primitives

### PR DESCRIPTION
**Pull request by JuaniRios**
## What

- Add the `st0x-execution` crate groundwork for extended-hours counter-trading: a `MarketSession` (Regular/Extended/Closed) classification from the broker calendar, `place_limit_order` + `cancel_order` (-> `CancellationOutcome`) + `fetch_latest_trade_price` on the `Executor` trait, and the `LimitOrder` / `OrderPlacement` extensions.
- Represent broker order outcomes faithfully: `OrderState::PartiallyFilled` / `Cancelled`, `OrderUpdate.shares_filled`, and `client_order_id` idempotency; parse `filled_qty` and map broker `PartiallyFilled` / `Canceled` statuses instead of collapsing them; drive status updates from broker timestamps (`filled_at` / `canceled_at`) with a terminal-timestamp fallback.
- `OrderPlacement` now records `extended_hours` and `limit_price` -- on a duplicate-`client_order_id` 422 the *adopted* order's terms are reported (for both market and limit placements) so a later convergence sweep sees broker reality, falling back to the request's `extended_hours` / `limit_price` only when the broker lookup omits them (market adoption defaults an omitted `extended_hours` to `false`).
- Adapt the main crate at the execution-API boundary so the workspace still compiles: `poll_status` dispatch, the trading CLI (`place_alpaca_limit_order` + new `OrderState` arms), the telemetry `Executor` decorator, and the executor test mocks.

## Why

- First implementation PR of the extended-hours counter-trading stack (#915 -> **#916** -> #931 -> #932). Isolating the broker-facing primitives keeps the execution layer independently reviewable before the `OffchainOrder` aggregate (#931) and the hedge/sweep wiring (#932) build on it.

## How

- All of `crates/execution/**` at the feature version. `Executor` gains `place_limit_order` / `cancel_order` as newly required methods; `market_session` / `fetch_latest_trade_price` have default impls.
- Boundary glue is deliberately conservative and behaviour-preserving (finalized in #931/#932): `poll_status` re-polls `PartiallyFilled` exactly as the pre-feature `Submitted` path did and treats a broker `Cancelled` as a rejection; the CLI renders/handles the new states and the test mocks forward the new trait methods.

## Testing

- New execution-crate unit tests cover the added states, `filled_qty` parsing, the broker-timestamp status mapping, and `client_order_id` idempotency; the boundary-glue `poll_status` / `cli::trading` tests are updated. Verified locally on the stack tip: `cargo check --all-targets`, `cargo clippy --all-features`, `cargo fmt --check` clean. CI runs the full suite per PR.

## Anything else

- Breaking within the execution crate: `OrderState::Filled` price `Float` -> `Usd`, `order_id` `String` -> `ExecutorOrderId`, `OrderState::Failed` gains `shares_filled` / `avg_price`, `AlpacaLimitOrder` now requires `client_order_id`, `IncompleteFilledOrder` -> `IncompleteOrder`, and the Alpaca-specific helper `place_limit_order` is renamed `place_alpaca_limit_order` (the trait's new `place_limit_order` takes the broker-agnostic `LimitOrder` and owns that name).
- Stack: #915 (groundwork) -> **#916** (execution) -> #931 (aggregate) -> #932 (rest). The lower PRs carry conservative placeholder handling that the upper PRs replace; the feature ships disabled.

### CODE REVIEW

**Comment by JuaniRios** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/916?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#932** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/932?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#917](https://github.com/ST0x-Technology/st0x.liquidity/pull/917) <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/917?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#941](https://github.com/ST0x-Technology/st0x.liquidity/pull/941) <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/941?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#931** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/931?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#916** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/916?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/916?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#915** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/915?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`
<hr />

## How to use the Graphite Merge Queue

Add the label <kbd>add-to-gt-merge-queue</kbd> to this PR to add it to the merge queue.

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/ST0x-Technology?ref=merge-queue-instructions-in-stack-comment&prId=916).

<h2></h2>
<sub>

An organization admin has required the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Comment by JuaniRios** [comment]
@coderabbitai full review

File: crates/execution/src/alpaca_broker_api/market_hours.rs [hunk 1]
```diff
@@ -32,33 +60,113 @@ pub(super) async fn is_market_open(
...
+/// Returns the market session at the given time.
+async fn market_session_at(
     client: &AlpacaBrokerApiClient,
     now: DateTime<Utc>,
-) -> Result<bool, AlpacaBrokerApiError> {
+) -> Result<MarketSession, AlpacaBrokerApiError> {
     let now_et = now.with_timezone(&New_York);
     let today = now_et.date_naive();
 
     let calendar = get_calendar(client, today, today).await?;
```
[the remark this answers is not in the document]
Comment (JuaniRios) [reply crates/execution/src/alpaca_broker_api/market_hours.rs]: Deferring this one. A per-day calendar cache breaks the `market_hours_transitions` e2e test, which drives session changes by mutating the broker calendar within a single day; the cache — correctly assuming the calendar is static per day in production — then never observes the simulated open, so the bot never hedges the pending position. A safe version needs either an injectable clock in the e2e harness or a cache that is invalidated/short-lived around session boundaries (and that only caches positive trading-day responses, so a transient empty calendar is not pinned for the whole day). Reverted here and tracking it as a follow-up rather than weaken the transition test.

**Comment by JuaniRios** [comment]
@coderabbitai full review

File: crates/execution/src/alpaca_broker_api/executor.rs [hunk 2]
```diff
@@ -226,6 +271,52 @@ impl Executor for AlpacaBrokerApi {
...
+        )
+        .await?;
+        Ok(Some(price))
+    }
+
+    async fn market_session(&self) -> Result<MarketSession, Self::Error> {
+        super::market_hours::market_session(&self.client).await
+    }
+
+    async fn place_limit_order(
```
Comment (0xgleb) [inline crates/execution/src/alpaca_broker_api/executor.rs:292]: should add coverage: this is the new `Executor::place_limit_order` for AlpacaBrokerApi, and it is the production entry point automated extended-hours hedging will call. It carries real glue that nothing exercises: the `AlpacaLimitPrice::try_new(order.limit_price)?` conversion and the broker-agnostic `LimitOrder` -> `AlpacaLimitOrder` field map. The three limit-placement tests in this file (test_place_limit_order_succeeds_for_active_tradable_asset and friends) were all re-pointed to the renamed inherent `place_alpaca_limit_order`, which takes a pre-built `AlpacaLimitPrice` and never touches this mapping or the try_new wiring. A grep confirms `.place_limit_order(LimitOrder { .. })` is called nowhere outside the definition. Most field mis-maps wouldn't compile, but a flipped/hardcoded extended_hours or a swapped price would ship untested, and the precision-rejection branch (try_new erroring on >2-decimal prices) is only reachable through this method for the automated path. Add one httpmock test that calls `executor.place_limit_order(LimitOrder { .. })`, asserts the serialized request (symbol/qty/side/limit_price/extended_hours/client_order_id) and the returned OrderPlacement, plus one feeding a >2-decimal price and asserting the precision error surfaces through this method.

File: crates/execution/src/alpaca_broker_api/market_hours.rs [hunk 3]
```diff
@@ -32,33 +62,116 @@ pub(super) async fn is_market_open(
...
+             extended-hours close (expected on early-close days)"
+        );
+    }
+
     let now_time = now_et.time();
-    let is_open = now_time >= today_calendar.open && now_time < today_calendar.close;
+
+    let session = if now_time >= today_calendar.open && now_time < today_calendar.close {
+        MarketSession::Regular
+    } else if now_time >= today_calendar.session_open && now_time < today_calendar.session_close {
```
Comment (0xgleb) [inline crates/execution/src/alpaca_broker_api/market_hours.rs:146]: minor: classification uses strict `now_time < today_calendar.session_close`, so exactly 20:00 ET must be Closed, but no test pins that instant. You pinned the other three transitions exactly (4:00 -> Extended, 9:30 -> Regular, 16:00 -> Extended); the 20:00 session_close edge is the only one left to a non-exact time (the nearest is the 21:00 case). An off-by-one here (accidentally `<=`) would classify 20:00 as Extended and place an extended-hours limit order the broker rejects, with no test to catch it. Add a `market_session_closed_at_session_close_boundary` test at exactly 20:00 ET asserting Closed, mirroring the existing exact-boundary tests.

File: src/cli/trading.rs [hunk 4]
```diff
@@ -500,7 +510,12 @@ async fn place_market_order_until_filled<Exec: Executor, W: Write>(
             OrderState::Failed { error_reason, .. } => {
                 anyhow::bail!("buy order failed: {error_reason:?}");
             }
-            OrderState::Pending | OrderState::Submitted { .. } => {
+            OrderState::Cancelled { .. } => {
+                anyhow::bail!("buy order was cancelled by the broker");
+            }
+            OrderState::PartiallyFilled { .. }
```
Comment (0xgleb) [inline src/cli/trading.rs:539]: minor: the new `PartiallyFilled` continue-polling arm in place_market_order_until_filled (grouped with Pending|Submitted) and the new PartiallyFilled/Cancelled display arms in order_status_command have no test. Only the new `Cancelled` bail branch is covered (market_buy_fails_when_the_broker_cancels_the_order). The poll-continuation is behavior, not just display: a misplaced arm could turn a partial fill into a hang or a false failure on a money-handling CLI path. Worth a test that drives place_market_order_until_filled with a MockExecutor returning PartiallyFilled and asserts it keeps polling rather than terminating; the display arms are lower value but easy to cover too.

File: crates/execution/src/order/state.rs [hunk 5]
```diff
@@ -1,116 +1,70 @@
...
         executed_at: DateTime<Utc>,
-        order_id: String,
-        price: Float,
+        order_id: ExecutorOrderId,
+        price: Usd,
+    },
+    Cancelled {
+        cancelled_at: DateTime<Utc>,
+        order_id: ExecutorOrderId,
+        shares_filled: Option<FractionalShares>,
```
Comment (0xgleb) [inline crates/execution/src/order/state.rs]: nit: `Cancelled.shares_filled` is `Option<FractionalShares>` whereas the analogous `PartiallyFilled.shares_filled` is a plain `FractionalShares`. The Alpaca executor's Cancelled arm already requires the field (ok_or_else -> MissingOrderField::FilledQty) and always wraps it `Some(..)`, so the real broker path never produces `None` (Alpaca reports filled_qty "0" when unfilled); only test/mock construction uses `None`. The Option therefore admits a state the producer refuses to emit and reintroduces the Some(0)-vs-None ambiguity that OrderUpdate::shares_filled has a doc warning about, without that warning here. Either make it a non-optional FractionalShares (callers use FractionalShares::new(float!(0)) for no fill, matching PartiallyFilled), or if the Option must stay, document the Some(0)/None equivalence on the variant. Failed keeping the Option is justified (rejected payloads genuinely omit it), but Cancelled has PartiallyFilled's required semantics.

File: crates/execution/src/alpaca_broker_api/order.rs [hunk 6]
```diff
@@ -189,6 +191,37 @@ pub(super) struct OrderResponse {
...
+    /// adoption rationale as `extended_hours`.
+    #[serde(
+        default,
+        deserialize_with = "deserialize_option_float_from_number_or_string"
+    )]
+    pub limit_price: Option<Float>,
+    /// Broker-side timestamps from the documented order entity
+    /// (https://docs.alpaca.markets/reference/getorderforaccount). These are
+    /// the event times that flow into downstream `Position.last_updated` and
+    /// recency logic -- never substitute the local observation time for a
```
Comment (0xgleb) [inline crates/execution/src/alpaca_broker_api/order.rs]: nit: this field doc says 'never substitute the local observation time for a terminal state's timestamp', but OrderStatus::Failed is a terminal state and its arm goes through `broker_time_or_observation(response.failed_at.or(response.updated_at), order_id)`, which falls back to `Utc::now()` when both are absent. The asymmetry is intentional and well-explained at the call site (rejected/expired payloads are messy, so Failed falls back rather than wedging rejection handling), but the blanket wording reads as an absolute the code deliberately breaks, so a future maintainer could 'fix' the Failed arm to fail-closed and reintroduce the poll-forever wedge. Scope the sentence to the states it actually governs, e.g. 'never substitute observation time for a Filled or Cancelled terminal state's timestamp'. The broker_time_or_observation doc itself is fine since it only names filled_at/canceled_at.

File: crates/execution/src/alpaca_broker_api/market_hours.rs [hunk 7]
```diff
@@ -32,33 +62,116 @@ pub(super) async fn is_market_open(
...
+    // Detect a silent redefinition of the undocumented session bounds (see
+    // the CONTRACT RISK note on `CalendarDay`). A NARROWED window is the
+    // dangerous direction -- fills landing inside the assumed 04:00-20:00
+    // extended window but outside Alpaca's would classify Closed and sit
+    // unhedged with no broker rejection to surface it -- so warn loudly when
+    // the broker's bounds differ from the documented extended-hours window.
+    let expected_session_open = NaiveTime::from_hms_opt(4, 0, 0);
+    let expected_session_close = NaiveTime::from_hms_opt(20, 0, 0);
+    if Some(today_calendar.session_open) != expected_session_open {
+        warn!(
```
Comment (0xgleb) [inline crates/execution/src/alpaca_broker_api/market_hours.rs:122]: nit: the session-bound mismatch blocks (warn! when session_open != 04:00, debug! when session_close != 20:00) and a narrowed early-close calendar are never exercised; every market_hours test feeds the standard 04:00/09:30/16:00/20:00 windows. The CONTRACT RISK note on CalendarDay explicitly calls early-close days an expected real scenario several days a year. The classification operators (>=/<) are already covered with standard windows, so this is low value, but a single test with an early-close entry (open 09:30, close 13:00, session_close 17:00) asserting Extended just after 13:00 and Closed just after 17:00 would lock in the narrowed-window behavior the doc says to expect.

**Review by 0xgleb — CHANGES_REQUESTED** [review]

**Review by 0xgleb — APPROVED** [review]
